# Practical 1

### [Christian Cabrera](https://www.cst.cam.ac.uk/people/chc79), University

of Cambridge

### [Radzim Sendyka](https://www.cst.cam.ac.uk/people/rs2071), University

of Cambridge

### [Carl Henrik Ek](http://carlhenrik.com), University of Cambridge

### [Neil D. Lawrence](http://inverseprobability.com), University of

Cambridge

### 2024-11-05

**Abstract**: In this lab session we look at setting up a SQL server,
creating and populating a database, and making joins between different
tables.

$$
$$

::: {.cell .markdown}

<!-- Do not edit this file locally. -->
<!-- Do not edit this file locally. -->
<!---->
<!-- Do not edit this file locally. -->
<!-- Do not edit this file locally. -->
<!-- The last names to be defined. Should be defined entirely in terms of macros from above-->
<!--

-->

**The check Session for this Practical is 7th November 2024.**

-   This practical should prepare you for the course assessment. Ensure
    that you have a solid understanding of the material, with particular
    emphasis on the AWS database setup. You should be able to use the
    same database that you set up here for the final assessment.
-   In that assessment, you will work with datasets that require initial
    setup processes. **Start your work on dataset access and database
    setup early** to avoid being blocked from work on subsequent stages
    later.
-   Some tasks will require you to develop skills for searching for
    multiple solutions and experimenting with different approaches,
    which lecture content may not cover. This environment closely
    resembles real-world data science and software engineering
    challenges, where there might not be a unique correct solution.

## Brief History of the Cloud

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_cloud/includes/history-of-cloud.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_cloud/includes/history-of-cloud.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

In the early days of the internet, companies could make use of open
source software, such as the Apache web server and the MySQL database
for providing on line stores, or other capabilities. But as part of
that, they normally had to provide their own server farms. For example,
the earliest server for running PageRank from 1996 was hosted on custom
made hardware built in a case of Mega Blocks (see Figure ).

<img class="" src="https://mlatcl.github.io/advds/./slides/diagrams//cloud/The_first_Google_computer_at_Stanford.jpg" style="width:50%">

Figure: <i>The web search engine was hosted on custom built hardware.
Photo credit [Christian Heilmann on
Flickr](https://www.flickr.com/photos/11414938@N00/2821326488).</i>

By September 2000, Google operated 5000 PCs for searching and web
crawling, using the LINUX operating system.[1]

Cloud computing gives you access to computing at a similar scale, but on
an as-needed basis. AWS launched S3 cloud storage in March 2006 and the
Elastic Compute Cloud followed in August 2006. These services were
inspired by the challenges they had scaling their own web service (known
as Obidos) for running the world’s largest e-commerce site. Recent
market share estimates indicate that AWS still retains around a third of
the cloud infrastructure market.[2]

[1] See [infolab at
Stanford](http://infolab.stanford.edu/pub/voy/museum/pictures/display/0-4-Google.htm)
for more details.

[2] There are various reports that try to track market share, here’s one
from Gartner from June 2021:
<https://www.gartner.com/en/newsroom/press-releases/2021-06-28-gartner-says-worldwide-iaas-public-cloud-services-market-grew-40-7-percent-in-2020>.
It looks at revenue that suggests Amazon retains 40% with Microsoft
second largest on 20%.

## UK Housing Datasets

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_datasets/includes/uk-housing-data.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_datasets/includes/uk-housing-data.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

The UK Price Paid data for housing in dates back to 1995 and contains
millions of transactions. This database is available at the [gov.uk
site](https://www.gov.uk/government/statistical-data-sets/price-paid-data-downloads).
The total data is over 4 gigabytes in size and it is available in a
single file or in multiple files splitted by years and semester. For
example, the first part of the data for 2018 is stored at
<http://prod.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com/pp-2018-part1.csv>.
By applying the divide and conquer principle, we will download the
splitted data because these files is less than 100MB each which makes
them easier to manage.

Let’s download first the two files that contain the price paid data for
the transactions that took place in the year 1995:

In [2]:
import requests

In [ ]:
# Base URL where the dataset is stored
base_url = "http://prod.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com"
# Downloading part 1 from 1995
file_name_part_1 = "/pp-1995-part1.csv"
url = base_url + file_name_part_1
response = requests.get(url)
if response.status_code == 200:
  with open("." + file_name_part_1, "wb") as file:
    file.write(response.content)
# Downloading part 2 from 1995
file_name_part_2 = "/pp-1995-part2.csv"
url = base_url + file_name_part_2
response = requests.get(url)
if response.status_code == 200:
  with open("." + file_name_part_2, "wb") as file:
    file.write(response.content)

The data is downloaded as CSV files in the files explorer of this
notebook. You can see that the two pieces of code that download the data
are quite similar. It makes sense to use a for loop to automate the way
we access the dataset for the different years. The following code will
download the data from 1996 to 2010.

In [ ]:
# Base URL where the dataset is stored
base_url = "http://prod.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com"
# File name with placeholders
file_name = "/pp-<year>-part<part>.csv"
for year in range(2007,2011):
  print ("Downloading data for year: " + str(year))
  for part in range(1,3):
    url = base_url + file_name.replace("<year>", str(year)).replace("<part>", str(part))
    response = requests.get(url)
    if response.status_code == 200:
      with open("." + file_name.replace("<year>", str(year)).replace("<part>", str(part)), "wb") as file:
        file.write(response.content)

If we think of reusability, it would be good to create a function that
can be called from anywhere in your code.

In [ ]:
import requests

In [1]:
def download_price_paid_data(year_from, year_to):
    # Base URL where the dataset is stored
    base_url = "http://prod.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com"
    """Download UK house price data for given year range"""
    # File name with placeholders
    file_name = "/pp-<year>-part<part>.csv"
    for year in range(year_from, (year_to+1)):
        print (f"Downloading data for year: {year}")
        for part in range(1,3):
            url = base_url + file_name.replace("<year>", str(year)).replace("<part>", str(part))
            response = requests.get(url)
            if response.status_code == 200:
                with open("." + file_name.replace("<year>", str(year)).replace("<part>", str(part)), "wb") as file:
                    file.write(response.content)

Now we can call the function to download the data between two given
years. For example, let’s download the data from 2011 to 2020 by calling
the defined function.

In [3]:
download_price_paid_data(2020, 2024)

UsageError: Line magic function `%sql` not found.


### Exercise 1

Add this function to your fynesse library and download the data from
2021 to 2024 using your library.

### Exercise 1 Answer

Write your answer to Exercise 1 here

In [ ]:
# Use this box for any code you need
import fynesse
%load_ext autoreload
%autoreload 2
fynesse.access.download_price_paid_data(2021, 2024)


## Cloud Hosted Database

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_access/includes/aws-database-setup.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_access/includes/aws-database-setup.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

The size of the data makes it unwieldy to manipulate directly in python
frameworks such as `pandas`. As a result we will host and access the
data in a *relational database*.

Using the following ideas: 1. A cloud hosted database (such as MariaDB
hosted on the AWS RDS service). 2. SQL code wrapped in appropriately
structured python functions. 3. Joining databases tables. You will
construct a database containing tables that contain all house prices,
latitudes and longitudes from the UK house price data base since 1995.

### Important Notes

**You will manipulate large datasets along this practical and the final
assessment. If your database is blocked or is not responsive after any
operation, have a look at the databases dashboard to see if it is using
too much CPU. Reboot the database if that is the case.** **If you
encounter problems with the online notebook (e.g., interrupted
connections with the AWS server), you can use a local IDE to work in
your machine.**

### SQL Database Server

A typical machine learning installation might have you running a
database from a cloud service (such as AWS, Azure or Google Cloud
Platform). That cloud service would host the database for you, and you
would pay according to the number of queries made. Popular SQL server
software includes \[`MariaDB`\] https://mariadb.org/) which is open
source, or [Microsoft’s SQL
Server](https://www.microsoft.com/en-gb/sql-server/sql-server-2019).

Many start-up companies were formed on the back of a `MySQL` server
hosted on top of AWS. Although since MySQL was sold to Sun, and then
passed on to Oracle, the open source community has turned its attention
to `MariaDB`, here’s the [AWS instructions on how to set up
`MariaDB`](https://aws.amazon.com/getting-started/hands-on/create-mariadb-db/).

If you were designing your own ride hailing app, or any other major
commercial software you would want to investigate whether you would need
to set up a central SQL server in one of these frameworks.

### Creating a MariaDB Server on AWS

In this section, we’ll review the setup required to create a MariaDB
server on AWS. The earliest AWS services of S3 and EC2 gave storage and
compute. Together these could be combined to host a database service.
Today cloud providers also provide machines that are already set up to
provide a database service. We will make use of AWS’s Relational
Database Service to provide our `MariaDB` server.

1.  Log in to your AWS account and go to the AWS RDS console
    [here](https://console.aws.amazon.com/rds/home).
2.  Make sure **the region is set to Europe (London) which is denoted as
    eu-west-2.**
3.  Scroll down to “Create Database”. Do *not* create an Aurora database
    instance.
4.  `Standard Create` should be selected. In the box below, which is
    titled `Engine Options` you should select `MariaDB`. You can leave
    the `Version` as it’s set.

<img class="" src="https://mlatcl.github.io/advds/./slides/diagrams//cloud/aws-select-mariadb-rds.png" style="width:60%">

Figure: <i>The AWS console box for selecting the `MariaDB` engine.</i>

1.  In the box below that, make sure you select `Free tier`.

<img class="" src="https://mlatcl.github.io/advds/./slides/diagrams//cloud/aws-select-free-tier.png" style="width:60%">

Figure: <i>Make sure you select the free tier option for your database
server.</i>

1.  Name your database server. For your setup we suggest you use
    `database-ads-<CRSid>` for the name. `<CRSid>` **corresponds to your
    CRSid. So, every student has an independent database.**
2.  Set a master password for accessing the database server as admin.

<img class="" src="https://mlatcl.github.io/advds/./slides/diagrams//cloud/aws-mariadb-settings.png" style="width:60%">

Figure: <i>Set the password and username for the database server
access.</i>

1.  Leave the `DB instance class` as it is.
2.  Leave the `DB instance size` at the default setting. Leave the
    storage type and allocated storage at the default settings of
    `General Purpose` (SSD) and `20`.
3.  *Disable* autoscaling in the `Storage Autoscaling` option.
4.  In the connectivity leave the VPC selection as `Default VPC` and
    *enable* `Publicly accessible` so that you’ll have an IP address for
    your database.
5.  In `VPC security group` select `Create new` to create a new security
    group for the instance.
6.  Write `ADSMariaDB` as the group name for the VPC security group.
7.  Leave the rest as it is and select `Create database` at the bottom
    to launch the database server.

Your database server will take a few minutes to launch.

While it’s launching you can check the access rules for the database
server
[here](https://eu-west-2.console.aws.amazon.com/ec2/v2/home?region=eu-west-2#SecurityGroups:).

1.  Select the `Default` security group.
2.  The source of the active inbound rule must be set to `0.0.0.0/0`. It
    means you can connect from any source using IPv4.

A wrong inbound rule can cause you fail connecting to the database from
this notebook.

*Note:* by setting the inbound rule to `0.0.0.0/0` we have opened up
access to *any* IP address. If this were production code you wouldn’t do
this, you would specify a range of addresses or the specific address of
the compute server that needed to access the system. Because we’re using
Google colab or another notebook client to access, and we can’t control
the IP address of that access, for simplicity we’ve set it up so that
any IP address can access the database, but that is *not good practice*
for production systems.

## Connecting to your Database Server

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_access/includes/sql-database-connection.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_access/includes/sql-database-connection.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

Before you start, you’re going to need the username and password you
set-up above for accessing the database server. You will need to make
use of it when your client connects to the server. It’s good practice to
never expose passwords in your code directly. So to protect your
passowrd, we’re going to create a `credentials.yaml` file locally that
will store your username and password so that the client can access the
server without ever showing your password in the notebook. This file
will also score the URL and port of your database. You can get these
details from your database connectivity and security details.

In [4]:
import yaml
from ipywidgets import interact_manual, Text, Password

In [14]:
@interact_manual(username=Text(description="Username:"),
                password=Password(description="Password:"),
                url=Text(description="URL:"),
                port=Text(description="Port:"))
def write_credentials(username, password, url, port):
    with open("credentials.yaml", "w") as file:
        credentials_dict = {'username': username,
                           'password': password,
                           'url': url,
                           'port': port}
        yaml.dump(credentials_dict, file)

interactive(children=(Text(value='', description='Username:'), Password(description='Password:'), Text(value='…

If you click `Run Interact` then the credentials you’ve selected will be
saved in the `yaml` file.

Then, we can read the server credentials using:

In [15]:
with open("credentials.yaml") as file:
  credentials = yaml.safe_load(file)
username = credentials["username"]
password = credentials["password"]
url = credentials["url"]
port = credentials["port"]

print(f"Username: {username}")
print(f"Password: {password}")
print(f"URL: {url}")
print(f"Port: {port}")

Username: admin
Password: hkyhank0813
URL: database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com
Port: 3306


## SQL Commands

We have all the required data to interact with our database server.
There are mainly two ways how we can do that. The first one is using
magic SQL commands. For this option, we need to install pymysql and load
the sql extension:

In [18]:
%pip install pymysql
%pip install ipython-sql


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.0/45.0 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 15.3 MB/s eta 0:00:00


In [19]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


We can now test our first database server connection using magic SQL.
The first line establishes the connection and the second one list the
databases. For now, you should see the databases that the engine has
installed by default.

In [ ]:
%pip install -U prettytable


In [20]:
%sql mariadb+pymysql://$username:$password@$url?local_infile=1
%sql SHOW databases

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
6 rows affected.


[('ads_2024',),
 ('information_schema',),
 ('innodb',),
 ('mysql',),
 ('performance_schema',),
 ('sys',)]

In [21]:
# -%sql KILL 675
%sql SHOW FULL PROCESSLIST

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
2 rows affected.


[(4, 'rdsadmin', 'localhost', 'mysql', 'Sleep', 4, '', None, 0.0),
 (7325, 'admin', '34.106.2.169:47746', None, 'Query', 0, 'starting', 'SHOW FULL PROCESSLIST', 0.0)]

This connection also enables the uploading of local files as part of the
connection (i.e., `local_infile=1`). We will use this property later.

## Database Schema

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_access/includes/sql-database-schema.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_access/includes/sql-database-schema.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

As a first step after establishing a connection, we should create our
own database in the server. We will use this relational database to
store, structure, and access the different datasets we will manipulate
during this course. We will use SQL code for this and once again magic
commands:

In [22]:
%%sql
SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
SET time_zone = "+00:00";

CREATE DATABASE IF NOT EXISTS `ads_2024` DEFAULT CHARACTER SET utf8 COLLATE utf8_bin;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.
0 rows affected.


[]

In [ ]:
%sql SHOW databases

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
6 rows affected.


[('ads_2024',),
 ('information_schema',),
 ('innodb',),
 ('mysql',),
 ('performance_schema',),
 ('sys',)]

After the database is created in our server, we must tell it which
database we will use. You will need to run this command after you create
a new connection:

In [23]:
%%sql
USE `ads_2024`;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.


[]

A database is composed of tables where data records are stored in rows.
The attributes of each record are the columns. We must define an
`schema` to create a table in a database. The `schema` tells the
database and the server what to expect in the columns of the table
(i.e., names and data types of the columns).

The `schema` is defined by the data we want to store. If we want to
store the UK Price Paid data, we should have a look at [its description
first](https://www.gov.uk/guidance/about-the-price-paid-data#explanations-of-column-headers-in-the-ppd).
Also, we should a look at the data. Let’s do that for the second
semester of 2021.

In [24]:
import pandas as pd

In [ ]:
file_name = "./pp-2021-part2.csv"
data = pd.read_csv(file_name)
data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 635765 entries, 0 to 635764
Data columns (total 16 columns):
 #   Column                                  Non-Null Count   Dtype 
---  ------                                  --------------   ----- 
 0   {C8A3A576-E3AE-0425-E053-6C04A8C0947B}  635765 non-null  object
 1   128065                                  635765 non-null  int64 
 2   2021-07-12 00:00                        635765 non-null  object
 3   CW12 2GS                                633303 non-null  object
 4   T                                       635765 non-null  object
 5   Y                                       635765 non-null  object
 6   F                                       635765 non-null  object
 7   8                                       635765 non-null  object
 8   Unnamed: 8                              79512 non-null   object
 9   CARTWRIGHT CLOSE                        622172 non-null  object
 10  EATON                                   246462 non-null 

Based on the columns of the dataframe, we should define now the
equivalent `schema` of the table in the SQL database. We will use once
more SQL magic commands to create a table with the equivalent `schema`:

In [25]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [26]:
%%sql
USE `ads_2024`;
DROP TABLE IF EXISTS `pp_data`;
CREATE TABLE IF NOT EXISTS `pp_data` (
  `transaction_unique_identifier` tinytext COLLATE utf8_bin NOT NULL,
  `price` int(10) unsigned NOT NULL,
  `date_of_transfer` date NOT NULL,
  `postcode` varchar(8) COLLATE utf8_bin NOT NULL,
  `property_type` varchar(1) COLLATE utf8_bin NOT NULL,
  `new_build_flag` varchar(1) COLLATE utf8_bin NOT NULL,
  `tenure_type` varchar(1) COLLATE utf8_bin NOT NULL,
  `primary_addressable_object_name` tinytext COLLATE utf8_bin NOT NULL,
  `secondary_addressable_object_name` tinytext COLLATE utf8_bin NOT NULL,
  `street` tinytext COLLATE utf8_bin NOT NULL,
  `locality` tinytext COLLATE utf8_bin NOT NULL,
  `town_city` tinytext COLLATE utf8_bin NOT NULL,
  `district` tinytext COLLATE utf8_bin NOT NULL,
  `county` tinytext COLLATE utf8_bin NOT NULL,
  `ppd_category_type` varchar(2) COLLATE utf8_bin NOT NULL,
  `record_status` varchar(2) COLLATE utf8_bin NOT NULL,
  `db_id` bigint(20) unsigned NOT NULL
) DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=1 ;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.
0 rows affected.


[]

The `schema` defines an id field in the table (i.e., `db_id`), which
must be unique and will play the role of [primary
key](https://www.geeksforgeeks.org/primary-key-in-dbms/), which is a
crucial concept in relational databases. The following code sets up the
primary key for our table and makes it auto increment when a new row
(i.e., record) is insterted into the table.

In [27]:
%%sql
--
-- Primary key for table `pp_data`
--
ALTER TABLE `pp_data`
ADD PRIMARY KEY (`db_id`);

ALTER TABLE `pp_data`
MODIFY db_id bigint(20) unsigned NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=1;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

In [28]:
%%sql
DESCRIBE pp_data;


 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
17 rows affected.


[('transaction_unique_identifier', 'tinytext', 'NO', '', None, ''),
 ('price', 'int(10) unsigned', 'NO', '', None, ''),
 ('date_of_transfer', 'date', 'NO', '', None, ''),
 ('postcode', 'varchar(8)', 'NO', '', None, ''),
 ('property_type', 'varchar(1)', 'NO', '', None, ''),
 ('new_build_flag', 'varchar(1)', 'NO', '', None, ''),
 ('tenure_type', 'varchar(1)', 'NO', '', None, ''),
 ('primary_addressable_object_name', 'tinytext', 'NO', '', None, ''),
 ('secondary_addressable_object_name', 'tinytext', 'NO', '', None, ''),
 ('street', 'tinytext', 'NO', '', None, ''),
 ('locality', 'tinytext', 'NO', '', None, ''),
 ('town_city', 'tinytext', 'NO', '', None, ''),
 ('district', 'tinytext', 'NO', '', None, ''),
 ('county', 'tinytext', 'NO', '', None, ''),
 ('ppd_category_type', 'varchar(2)', 'NO', '', None, ''),
 ('record_status', 'varchar(2)', 'NO', '', None, ''),
 ('db_id', 'bigint(20) unsigned', 'NO', 'PRI', None, 'auto_increment')]

We’ve created our first table in the database with its respective
primary key. Now we need to populate it. There are different methods to
do that. [Some of them more efficient than
others](https://dev.to/arctype/load-data-infile-vs-insert-in-mysql-why-how-when-247f#:~:text=%E2%80%8BWhen%20working%20with%20MySQL,way%20faster%20than%20INSERT%20does.).
In our case, given the size of our data set, we will take advantage of
the csv files we downloaded in the first part of this lab. The command
`LOAD DATA LOCAL INFILE` allows uploading data to the table from a CSV
file. We must specify the name of the local file, the name of the table,
and the format of the CSV file we want to use (i.e., separators,
enclosers, termination line characters, etc.)

The following command uploads the data of the transactions that took
place in 1995.

In [ ]:
%sql USE `ads_2024`;
%sql LOAD DATA LOCAL INFILE './pp-1996-part1.csv' INTO TABLE `pp_data` FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED by '"' LINES STARTING BY '' TERMINATED BY '\n';
%sql LOAD DATA LOCAL INFILE './pp-1996-part2.csv' INTO TABLE `pp_data` FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED by '"' LINES STARTING BY '' TERMINATED BY '\n';

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
473709 rows affected.
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
473482 rows affected.


[]

If we want to upload the data for all the years, we will need a command
for each CSV in our dataset. Alternatively, we can use Python code
together with SQL magic commands as follows:

In [30]:
for year in range(2020, 2025):
    print("Uploading data for year: " + str(year))
    for part in range(1, 3):
        file_name = f'./pp-{year}-part{part}.csv'
        print(f"Loading data from {file_name}")
        %sql LOAD DATA LOCAL INFILE :file_name INTO TABLE `pp_data` FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '"' LINES TERMINATED BY '\n';

Uploading data for year: 2020
Loading data from ./pp-2020-part1.csv
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
446812 rows affected.
Loading data from ./pp-2020-part2.csv
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
446813 rows affected.
Uploading data for year: 2021
Loading data from ./pp-2021-part1.csv
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
635765 rows affected.
Loading data from ./pp-2021-part2.csv
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
635766 rows affected.
Uploading data for year: 2022
Loading data from ./pp-2022-part1.csv
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
528907 rows affected.
Loading data from ./pp-2022-part2.csv
 * mariadb+pymysql://admin:***@database-ad

In [ ]:
%%sql
--
-- Primary key for table `pp_data`
--
ALTER TABLE `pp_data`
ADD PRIMARY KEY (`db_id`);

ALTER TABLE `pp_data`
MODIFY db_id bigint(20) unsigned NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=1;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
(pymysql.err.OperationalError) (1068, 'Multiple primary key defined')
[SQL: --
-- Primary key for table `pp_data`
--
ALTER TABLE `pp_data`
ADD PRIMARY KEY (`db_id`);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


Now that we uploaded the data, we can retrieve it from the table. We can
select the first 5 elements in the `pp_data` using this command:

In [31]:
%%sql
USE `ads_2024`;
SELECT * FROM `pp_data` LIMIT 5;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
5 rows affected.


[('{BA558B33-0888-76EF-E053-6B04A8C0B4B7}', 193000, datetime.date(2020, 11, 2), 'IP27 9RD', 'T', 'N', 'F', '17', '', 'CHESTNUT WAY', 'RAF LAKENHEATH', 'BRANDON', 'WEST SUFFOLK', 'SUFFOLK', 'A', 'A', 1),
 ('{BA558B33-0889-76EF-E053-6B04A8C0B4B7}', 500000, datetime.date(2020, 11, 27), 'IP13 9FG', 'D', 'N', 'F', '7', '', 'CLARKE DRIVE', 'FRAMLINGHAM', 'WOODBRIDGE', 'EAST SUFFOLK', 'SUFFOLK', 'A', 'A', 2),
 ('{BA558B33-088C-76EF-E053-6B04A8C0B4B7}', 142000, datetime.date(2020, 12, 7), 'IP28 7JL', 'F', 'N', 'L', 'THE OLD VICARAGE, 35', 'FLAT 7', 'QUEENSWAY', 'MILDENHALL', 'BURY ST EDMUNDS', 'WEST SUFFOLK', 'SUFFOLK', 'A', 'A', 3),
 ('{BA558B33-088D-76EF-E053-6B04A8C0B4B7}', 142000, datetime.date(2020, 11, 27), 'IP33 3SH', 'F', 'N', 'L', '22', '', 'PRINCE OF WALES CLOSE', '', 'BURY ST EDMUNDS', 'WEST SUFFOLK', 'SUFFOLK', 'A', 'A', 4),
 ('{BA558B33-088E-76EF-E053-6B04A8C0B4B7}', 175000, datetime.date(2020, 11, 9), 'NR33 7HR', 'T', 'N', 'F', '163', '', 'STRADBROKE ROAD', '', 'LOWESTOFT', 'EAST SUFFOLK', 'SUFFOLK', 'A', 'A', 5)]

We can also count the number of rows in our table. It can take more than
5 minutes to finish. There are almost 30 million of records in the
dataset.

In [32]:
%sql select count(*) from `pp_data`;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
1 rows affected.


[(4370411,)]

## Postal Codes Dataset

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_datasets/includes/postcode-data.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_datasets/includes/postcode-data.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

The UK Price Paid dataset is now available in our database. This dataset
provides useful information about the housing transactions in the UK.
More complete analysis can be enabled if we join it with additional
datasets. That is the goal of the rest of this practical. The [Open
Postcode Geo](https://www.getthedata.com/open-postcode-geo) provides
additional information about the houses. It is a dataset of British
postcodes with easting, northing, latitude, and longitude and with
additional fields for geospace applications, including postcode area,
postcode district, postcode sector, incode, and outcode.

Your task now is to make this dataset available and accessible in our
database. The data you need can be found at this url:
<https://www.getthedata.com/downloads/open_postcode_geo.csv.zip>. It
will need to be unzipped before use.

You may find the following schema useful for the postcode data
(developed by Christian and Neil)

In [33]:
%%sql
--
-- Table structure for table `postcode_data`
--
DROP TABLE IF EXISTS `postcode_data`;
CREATE TABLE IF NOT EXISTS `postcode_data` (
  `postcode` varchar(8) COLLATE utf8_bin NOT NULL,
  `status` enum('live','terminated') NOT NULL,
  `usertype` enum('small', 'large') NOT NULL,
  `easting` int unsigned,
  `northing` int unsigned,
  `positional_quality_indicator` int NOT NULL,
  `country` enum('England', 'Wales', 'Scotland', 'Northern Ireland', 'Channel Islands', 'Isle of Man') NOT NULL,
  `latitude` decimal(11,8) NOT NULL,
  `longitude` decimal(10,8) NOT NULL,
  `postcode_no_space` tinytext COLLATE utf8_bin NOT NULL,
  `postcode_fixed_width_seven` varchar(7) COLLATE utf8_bin NOT NULL,
  `postcode_fixed_width_eight` varchar(8) COLLATE utf8_bin NOT NULL,
  `postcode_area` varchar(2) COLLATE utf8_bin NOT NULL,
  `postcode_district` varchar(4) COLLATE utf8_bin NOT NULL,
  `postcode_sector` varchar(6) COLLATE utf8_bin NOT NULL,
  `outcode` varchar(4) COLLATE utf8_bin NOT NULL,
  `incode` varchar(3)  COLLATE utf8_bin NOT NULL,
  `db_id` bigint(20) unsigned NOT NULL
) DEFAULT CHARSET=utf8 COLLATE=utf8_bin;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

And again you’ll want to set up a primary key for the new table.

In [36]:
%%sql
ALTER TABLE `postcode_data`
ADD PRIMARY KEY (`db_id`);

ALTER TABLE `postcode_data`
MODIFY `db_id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,AUTO_INCREMENT=1;



 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
(pymysql.err.OperationalError) (1068, 'Multiple primary key defined')
[SQL: ALTER TABLE `postcode_data`
ADD PRIMARY KEY (`db_id`);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


### Exercise 2

Upload the postcode dataset to your database using the LOAD DATA LOCAL
INFILE command.

### Exercise 2 Answer

Write your answer to Exercise 2 here

In [ ]:
# clearing the table

%sql USE `ads_2024`;
%sql TRUNCATE TABLE `postcode_data`;




 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.


[]

In [37]:
%sql USE `ads_2024`;
%sql LOAD DATA LOCAL INFILE './open_postcode_geo.csv' INTO TABLE `postcode_data` FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED by '"' LINES STARTING BY '' TERMINATED BY '\n';

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
(pymysql.err.OperationalError) (1017, "Can't find file 'b'./open_postcode_geo.csv''")
[SQL: LOAD DATA LOCAL INFILE './open_postcode_geo.csv' INTO TABLE `postcode_data` FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED by '"' LINES STARTING BY '' TERMINATED BY '\n' ;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [ ]:
%sql DESCRIBE postcode_data;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
18 rows affected.


[('postcode', 'varchar(8)', 'NO', '', None, ''),
 ('status', "enum('live','terminated')", 'NO', '', None, ''),
 ('usertype', "enum('small','large')", 'NO', '', None, ''),
 ('easting', 'int(10) unsigned', 'YES', '', None, ''),
 ('northing', 'int(10) unsigned', 'YES', '', None, ''),
 ('positional_quality_indicator', 'int(11)', 'NO', '', None, ''),
 ('country', "enum('England','Wales','Scotland','Northern Ireland','Channel Islands','Isle of Man')", 'NO', '', None, ''),
 ('latitude', 'decimal(11,8)', 'NO', '', None, ''),
 ('longitude', 'decimal(10,8)', 'NO', '', None, ''),
 ('postcode_no_space', 'tinytext', 'NO', '', None, ''),
 ('postcode_fixed_width_seven', 'varchar(7)', 'NO', '', None, ''),
 ('postcode_fixed_width_eight', 'varchar(8)', 'NO', '', None, ''),
 ('postcode_area', 'varchar(2)', 'NO', '', None, ''),
 ('postcode_district', 'varchar(4)', 'NO', '', None, ''),
 ('postcode_sector', 'varchar(6)', 'NO', '', None, ''),
 ('outcode', 'varchar(4)', 'NO', '', None, ''),
 ('incode', 'varchar(3)', 'NO', '', None, ''),
 ('db_id', 'bigint(20) unsigned', 'NO', 'PRI', None, 'auto_increment')]

In [ ]:
%%sql
SELECT * FROM `postcode_data` LIMIT 5;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
5 rows affected.


[('AB1 0AA', 'terminated', 'small', 385386, 801193, 1, 'Scotland', Decimal('57.10147400'), Decimal('-2.24285100'), 'AB10AA', 'AB1 0AA', 'AB1  0AA', 'AB', 'AB1', 'AB1 0', 'AB1', '0AA', 1),
 ('AB1 0AB', 'terminated', 'small', 385177, 801314, 1, 'Scotland', Decimal('57.10255400'), Decimal('-2.24630800'), 'AB10AB', 'AB1 0AB', 'AB1  0AB', 'AB', 'AB1', 'AB1 0', 'AB1', '0AB', 2),
 ('AB1 0AD', 'terminated', 'small', 385053, 801092, 1, 'Scotland', Decimal('57.10055600'), Decimal('-2.24834200'), 'AB10AD', 'AB1 0AD', 'AB1  0AD', 'AB', 'AB1', 'AB1 0', 'AB1', '0AD', 3),
 ('AB1 0AE', 'terminated', 'small', 384600, 799300, 8, 'Scotland', Decimal('57.08444400'), Decimal('-2.25570800'), 'AB10AE', 'AB1 0AE', 'AB1  0AE', 'AB', 'AB1', 'AB1 0', 'AB1', '0AE', 4),
 ('AB1 0AF', 'terminated', 'large', 384460, 800660, 8, 'Scotland', Decimal('57.09665600'), Decimal('-2.25810200'), 'AB10AF', 'AB1 0AF', 'AB1  0AF', 'AB', 'AB1', 'AB1 0', 'AB1', '0AF', 5)]

## Joining Tables

<span class="editsection-bracket"
style="">\[</span><span class="editsection"
style=""><a href="https://github.com/lawrennd/snippets/edit/main/_access/includes/table-joins.md" target="_blank" onclick="ga('send', 'event', 'Edit Page', 'Edit', 'https://github.com/lawrennd/snippets/edit/main/_access/includes/table-joins.md', 13);">edit</a></span><span class="editsection-bracket" style="">\]</span>

When we have two tables with data tha describe the same object, then it
makes sense to join the tables together to enrich our knowledge about
the object. For example, while the `pp_data` tell us about the
transactions a house have been involved, the `postcode_data` tell us
details about the location of the house. By joining both, we could
answer more interesting questions like what are the coordinates of the
most expensive house in 2024?

The join of the tables must be done by matching the columns the tables
share. In this case, the `pp_data` and the `postcode_data` tables share
the column `postcode`. This operation can take long time because the
number of records in each database is huge.

You will find the issue of operations taking too long when handling
large data sets in different scenarios. An appropriate strategy to
overcome such issues is to index the tables. Indexing is a way to
organise the data so the queries are more efficient time-wise. Now the
task is to select the right columns to create the index. This decision
depends on the columns we are using in the SQL operations. In the join
case, we are selecting and matching records in each table using the
`postcode` column. It makes sense then to create an index for these
columns. The following code indexes the table `pp_data` by `postcode`.

In [ ]:
%%sql
USE `ads_2024`;
CREATE INDEX idx_pp_postcode ON pp_data(postcode);

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
(pymysql.err.OperationalError) (1061, "Duplicate key name 'idx_pp_postcode'")
[SQL: CREATE INDEX idx_pp_postcode ON pp_data(postcode);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [ ]:
%sql SHOW INDEX FROM pp_data;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
2 rows affected.


[('pp_data', 1, 'idx_pp_postcode', 1, 'postcode', 'A', 1120825, None, None, '', 'BTREE', '', '', 'NO'),
 ('pp_data', 1, 'idx_pp_date_transfer', 1, 'date_of_transfer', 'A', 1, None, None, '', 'BTREE', '', '', 'NO')]

In [ ]:
%%sql
USE `ads_2024`;
SHOW TABLES;


 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
3 rows affected.


[('postcode_data',), ('pp_data',), ('prices_coordinates_data',)]

Let’s try our join the tables for year 2024 now:

In [ ]:
%%sql
USE `ads_2024`;
select * from pp_data as pp inner join postcode_data as po on pp.postcode = po.postcode where pp.date_of_transfer BETWEEN '2024-01-01' AND '2024-12-31' limit 5;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

### Exercise 3

The index made a difference in the time the join operation took to
finish. Write the code to index the table `postcode_data` by `postcode`.

### Exercise 3 Answer

Write your answer to Exercise 3 here

In [ ]:

%%sql
USE `ads_2024`;
CREATE INDEX idx_postcode_data_postcode ON postcode_data(postcode);


 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

In [ ]:
%sql SHOW INDEX FROM postcode_data;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
2 rows affected.


[('postcode_data', 0, 'PRIMARY', 1, 'db_id', 'A', 2319747, None, None, '', 'BTREE', '', '', 'NO'),
 ('postcode_data', 1, 'idx_postcode_data_postcode', 1, 'postcode', 'A', 2319747, None, None, '', 'BTREE', '', '', 'NO')]

And, let’s try the join again:

In [ ]:
%%sql
USE `ads_2024`;
select * from pp_data as pp inner join postcode_data as po on pp.postcode = po.postcode where pp.date_of_transfer BETWEEN '2024-01-01' AND '2024-12-31' limit 5;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

### Exercise 4

Do you see any difference after adding the new index? Why?

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.


[]

In [ ]:
%%sql
USE `ads_2024`;
SELECT * FROM pp_data
WHERE date_of_transfer BETWEEN '2024-01-01' AND '2024-12-31'
LIMIT 5;


 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.


KeyboardInterrupt: 

### Exercise 4 Answer

Write your answer to Exercise 4 here

# Use this box for any code you need
Adding an index on the postcode column in both pp_data and postcode_data tables improves performance. This is because the database can quickly locate matching rows using the index, making joins faster. Queries that filter by postcode are faster, as the index allows direct access to matching rows.

## Database Python Client

Now let’s focus on the second way of interacting with the database
server. We can use Python code to create a client that communicates with
our database server. For this, we need to install the following
libraries.

In [ ]:
%pip install ipython-sql

Note: you may need to restart the kernel to use updated packages.


In [ ]:
%pip install PyMySQL

Note: you may need to restart the kernel to use updated packages.


Let’s create a method in Python to establish a database connection
wherever we like. It should look like the following code:

In [ ]:
import pymysql

In [ ]:
def create_connection(user, password, host, database, port=3306):
    """ Create a database connection to the MariaDB database
        specified by the host url and database name.
    :param user: username
    :param password: password
    :param host: host url
    :param database: database name
    :param port: port number
    :return: Connection object or None
    """
    conn = None
    try:
        conn = pymysql.connect(user=user,
                               passwd=password,
                               host=host,
                               port=port,
                               local_infile=1,
                               db=database
                               )
        print(f"Connection established!")
    except Exception as e:
        print(f"Error connecting to the MariaDB Server: {e}")
    return conn

Please add the code above to your fynesse library. We can now call this
function to get a connection:

In [ ]:
#Write your code to establish a connection using your fynesee library

Now let’s define a Python method that uploads to a table the data
product of the join operation between the tables `pp_data` and
`postcode_data`. For this, we first need to create the table that will
store this data.

In [ ]:
%%sql
USE `ads_2024`;
--
-- Table structure for table `prices_coordinates_data`
--
DROP TABLE IF EXISTS `prices_coordinates_data`;
CREATE TABLE IF NOT EXISTS `prices_coordinates_data` (
  `price` int(10) unsigned NOT NULL,
  `date_of_transfer` date NOT NULL,
  `postcode` varchar(8) COLLATE utf8_bin NOT NULL,
  `property_type` varchar(1) COLLATE utf8_bin NOT NULL,
  `new_build_flag` varchar(1) COLLATE utf8_bin NOT NULL,
  `tenure_type` varchar(1) COLLATE utf8_bin NOT NULL,
  `locality` tinytext COLLATE utf8_bin NOT NULL,
  `town_city` tinytext COLLATE utf8_bin NOT NULL,
  `district` tinytext COLLATE utf8_bin NOT NULL,
  `county` tinytext COLLATE utf8_bin NOT NULL,
  `country` enum('England', 'Wales', 'Scotland', 'Northern Ireland', 'Channel Islands', 'Isle of Man') NOT NULL,
  `latitude` decimal(11,8) NOT NULL,
  `longitude` decimal(10,8) NOT NULL,
  `db_id` bigint(20) unsigned NOT NULL
) DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=1 ;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.
0 rows affected.


[]

We should define the primary key for this table too.

In [ ]:
%%sql
ALTER TABLE `prices_coordinates_data`
ADD PRIMARY KEY (`db_id`);

ALTER TABLE `prices_coordinates_data`
MODIFY `db_id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,AUTO_INCREMENT=1;

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

Indexing the table `pp_data` by date will be useful for populating the
`prices_coordinates_data`. This index can take around 8 minutes to
finish.

In [ ]:
%%sql
USE `ads_2024`;
CREATE INDEX idx_pp_date_transfer ON pp_data(date_of_transfer);

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
0 rows affected.


[]

Now that the table exists in our database, let’s create a method for
uploading the join data. This method will upload the data for a given
year and will use the logic we have used before but in Python code.

In [ ]:
import csv

In [ ]:
import time

In [ ]:
def housing_upload_join_data(conn, year):
  start_date = str(year) + "-01-01"
  end_date = str(year) + "-12-31"

  cur = conn.cursor()
  print('Selecting data for year: ' + str(year))
  cur.execute(f'SELECT pp.price, pp.date_of_transfer, po.postcode, pp.property_type, pp.new_build_flag, pp.tenure_type, pp.locality, pp.town_city, pp.district, pp.county, po.country, po.latitude, po.longitude FROM (SELECT price, date_of_transfer, postcode, property_type, new_build_flag, tenure_type, locality, town_city, district, county FROM pp_data WHERE date_of_transfer BETWEEN "' + start_date + '" AND "' + end_date + '") AS pp INNER JOIN postcode_data AS po ON pp.postcode = po.postcode')
  rows = cur.fetchall()

  csv_file_path = 'output_file.csv'

  # Write the rows to the CSV file
  with open(csv_file_path, 'w', newline='') as csvfile:
    csv_writer = csv.writer(csvfile)
    # Write the data rows
    csv_writer.writerows(rows)
  print('Storing data for year: ' + str(year))
  cur.execute(f"LOAD DATA LOCAL INFILE '" + csv_file_path + "' INTO TABLE `prices_coordinates_data` FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED by '\"' LINES STARTING BY '' TERMINATED BY '\n';")
  print('Data stored for year: ' + str(year))

Now, lets upload the joined data for 2024. This upload is going to take
long time given the size of our datasets:

In [ ]:
conn = create_connection(username, password, url, "ads_2024", 3306)
housing_upload_join_data(conn, 2024)


Connection established!
Selecting data for year: 2024
Storing data for year: 2024
Data stored for year: 2024


### Exercise 5

Add the `housing_upload_join_data` function to your fynesse library and
write the code to upload the joined data for 2023.

### Exercise 5 Answer

Write your answer to Exercise 5 here

In [ ]:
# Use this box for any code you need

housing_upload_join_data(conn, 2023)

Selecting data for year: 2023
Storing data for year: 2023
Data stored for year: 2023


In [ ]:
%sql SHOW TABLES

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
3 rows affected.


[('postcode_data',), ('pp_data',), ('prices_coordinates_data',)]

To finalise this lab, let’s have a look at the structure of your
database running the following code:

In [ ]:
tables = %sql SHOW TABLES;

for row in tables:
    table_name = row[0]
    print(f"\nTable: {table_name}")

    table_status = %sql SHOW TABLE STATUS LIKE '{table_name}';
    approx_row_count = table_status[0][4] if table_status else 'Unable to fetch row count'
    print("\nApprox Row Count:", approx_row_count//100000/10, "M")

    first_5_rows = %sql SELECT * FROM `{table_name}` LIMIT 5;
    print(first_5_rows)

    indices = %sql SHOW INDEX FROM `{table_name}`;
    if indices:
        print("\nIndices:")
        for index in indices:
            print(f" - {index[2]} ({index[10]}): Column {index[4]}")
    else:
        print("\nNo indices set on this table.")

 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
3 rows affected.

Table: postcode_data
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
1 rows affected.

Approx Row Count: 2.3 M
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
5 rows affected.
+----------+------------+----------+---------+----------+------------------------------+----------+-------------+-------------+-------------------+----------------------------+----------------------------+---------------+-------------------+-----------------+---------+--------+-------+
| postcode |   status   | usertype | easting | northing | positional_quality_indicator | country  |   latitude  |  longitude  | postcode_no_space | postcode_fixed_width_seven | postcode_fixed_width_eight | postcode_area | postcode_district | postcode_sector | outcode | incode | db_id |
+---------

### Exercise 6

Write the output of the above code:

### Exercise 6 Answer

Write your answer to Exercise 6 here

# Use this box for any code you need
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
3 rows affected.

Table: postcode_data
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
1 rows affected.

Approx Row Count: 2.3 M
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
5 rows affected.
+----------+------------+----------+---------+----------+------------------------------+----------+-------------+-------------+-------------------+----------------------------+----------------------------+---------------+-------------------+-----------------+---------+--------+-------+
| postcode |   status   | usertype | easting | northing | positional_quality_indicator | country  |   latitude  |  longitude  | postcode_no_space | postcode_fixed_width_seven | postcode_fixed_width_eight | postcode_area | postcode_district | postcode_sector | outcode | incode | db_id |
+----------+------------+----------+---------+----------+------------------------------+----------+-------------+-------------+-------------------+----------------------------+----------------------------+---------------+-------------------+-----------------+---------+--------+-------+
| AB1 0AA  | terminated |  small   |  385386 |  801193  |              1               | Scotland | 57.10147400 | -2.24285100 |       AB10AA      |          AB1 0AA           |          AB1  0AA          |       AB      |        AB1        |      AB1 0      |   AB1   |  0AA   |   1   |
| AB1 0AB  | terminated |  small   |  385177 |  801314  |              1               | Scotland | 57.10255400 | -2.24630800 |       AB10AB      |          AB1 0AB           |          AB1  0AB          |       AB      |        AB1        |      AB1 0      |   AB1   |  0AB   |   2   |
| AB1 0AD  | terminated |  small   |  385053 |  801092  |              1               | Scotland | 57.10055600 | -2.24834200 |       AB10AD      |          AB1 0AD           |          AB1  0AD          |       AB      |        AB1        |      AB1 0      |   AB1   |  0AD   |   3   |
| AB1 0AE  | terminated |  small   |  384600 |  799300  |              8               | Scotland | 57.08444400 | -2.25570800 |       AB10AE      |          AB1 0AE           |          AB1  0AE          |       AB      |        AB1        |      AB1 0      |   AB1   |  0AE   |   4   |
| AB1 0AF  | terminated |  large   |  384460 |  800660  |              8               | Scotland | 57.09665600 | -2.25810200 |       AB10AF      |          AB1 0AF           |          AB1  0AF          |       AB      |        AB1        |      AB1 0      |   AB1   |  0AF   |   5   |
+----------+------------+----------+---------+----------+------------------------------+----------+-------------+-------------+-------------------+----------------------------+----------------------------+---------------+-------------------+-----------------+---------+--------+-------+
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
2 rows affected.

Indices:
 - PRIMARY (BTREE): Column db_id
 - idx_postcode_data_postcode (BTREE): Column postcode

Table: pp_data
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
1 rows affected.

Approx Row Count: 29.1 M
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
5 rows affected.
+------------------------------------------+-------+------------------+----------+---------------+----------------+-------------+---------------------------------+-----------------------------------+-------------------+-------------+-------------+------------------+------------------+-------------------+---------------+-------+
|      transaction_unique_identifier       | price | date_of_transfer | postcode | property_type | new_build_flag | tenure_type | primary_addressable_object_name | secondary_addressable_object_name |       street      |   locality  |  town_city  |     district     |      county      | ppd_category_type | record_status | db_id |
+------------------------------------------+-------+------------------+----------+---------------+----------------+-------------+---------------------------------+-----------------------------------+-------------------+-------------+-------------+------------------+------------------+-------------------+---------------+-------+
| "{B6F6A33A-B5BF-46FB-A2CD-BF5CDCD313E8}" |   0   |    0000-00-00    | "PE10 9T |       "       |       "        |      "      |              "11A"              |                 ""                |  "WETHERBY CLOSE" |   "BOURNE"  |   "BOURNE"  | "SOUTH KESTEVEN" |  "LINCOLNSHIRE"  |         "A        |       "A      |   0   |
| "{D83DD2D0-12B4-41B0-A113-BF5CE28823C4}" |   0   |    0000-00-00    | "LA2 6QQ |       "       |       "        |      "      |               "17"              |                 ""                |    "BEECH ROAD"   |   "HALTON"  | "LANCASTER" |   "LANCASTER"    |   "LANCASHIRE"   |         "A        |       "A      |   0   |
| "{05F4A2BD-7012-470D-AE2C-BF5CE477B766}" |   0   |    0000-00-00    | "E7 8EF" |       "       |       "        |      "      |               "48"              |                 ""                |  "HENDERSON ROAD" |   "LONDON"  |   "LONDON"  |     "NEWHAM"     | "GREATER LONDON" |         "A        |       "A      |   0   |
| "{02F7B636-12CE-441D-9EC4-B8486513A3B2}" |   0   |    0000-00-00    | "DE55 5J |       "       |       "        |      "      |              "274"              |                 ""                |  "ALFRETON ROAD"  | "BLACKWELL" |  "ALFRETON" |    "BOLSOVER"    |   "DERBYSHIRE"   |         "A        |       "A      |   0   |
| "{48490694-5C4B-48B8-B240-C2E3CB522E23}" |   0   |    0000-00-00    | "WF17 0E |       "       |       "        |      "      |               "29"              |                 ""                | "LANSDOWNE CLOSE" |   "BATLEY"  |   "BATLEY"  |    "KIRKLEES"    | "WEST YORKSHIRE" |         "A        |       "A      |   0   |
+------------------------------------------+-------+------------------+----------+---------------+----------------+-------------+---------------------------------+-----------------------------------+-------------------+-------------+-------------+------------------+------------------+-------------------+---------------+-------+
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
2 rows affected.

Indices:
 - idx_pp_postcode (BTREE): Column postcode
 - idx_pp_date_transfer (BTREE): Column date_of_transfer

Table: prices_coordinates_data
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
1 rows affected.

Approx Row Count: 0.0 M
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
0 rows affected.
+-------+------------------+----------+---------------+----------------+-------------+----------+-----------+----------+--------+---------+----------+-----------+-------+
| price | date_of_transfer | postcode | property_type | new_build_flag | tenure_type | locality | town_city | district | county | country | latitude | longitude | db_id |
+-------+------------------+----------+---------------+----------------+-------------+----------+-----------+----------+--------+---------+----------+-----------+-------+
+-------+------------------+----------+---------------+----------------+-------------+----------+-----------+----------+--------+---------+----------+-----------+-------+
 * mariadb+pymysql://admin:***@database-ads-kh747.cgrre17yxw11.eu-west-2.rds.amazonaws.com?local_infile=1
1 rows affected.

Indices:
 - PRIMARY (BTREE): Column db_id


## Summary

In this practical, we have explored how to persist a couple of datasets
in a relational database to facilitate future access. We configured a
Cloud-hosted database server and had a look at two ways to interact with
it. Then, we explored how to join tables using SQL and the benefits of
indexing. The tables you created in this practical will be used along
the course and we expect you use them for your final assignment. In the
following practical, you will `assess` this data using different methods
from visualisations to statistical analysis.

## Thanks!

For more information on these subjects and more you might want to check
the following resources.

-   book: [The Atomic
    Human](https://www.penguin.co.uk/books/455130/the-atomic-human-by-lawrence-neil-d/9780241625248)
-   twitter: [@lawrennd](https://twitter.com/lawrennd)
-   podcast: [The Talking Machines](http://thetalkingmachines.com)
-   newspaper: [Guardian Profile
    Page](http://www.theguardian.com/profile/neil-lawrence)
-   blog:
    [http://inverseprobability.com](http://inverseprobability.com/blog.html)

## References